reference-implementation/lib/readable-stream.js

```diff
@@ -98,57 +93,45 @@ export default class ReadableStream {
       resolvePipeToPromise = resolve;
       rejectPipeToPromise = reject;
 
-      source = this.getReader();
+      source.closed.catch(abortDest);
+      dest.closed.then(
+        () => {
+          if (!closedPurposefully) {
+            cancelSource(new TypeError('destination is closing or closed and cannot be piped to anymore'));
+          }
+        },
+        cancelSource
+      );
+
       doPipe();
     });
 
     function doPipe() {
-      for (;;) {
-        const ds = dest.state;
-        if (ds === 'writable') {
-          if (source.state === 'readable') {
-            dest.write(source.read());
-            continue;
-          } else if (source.state === 'waiting') {
-            Promise.race([source.ready, dest.closed]).then(doPipe, doPipe);
-          } else if (source.state === 'errored') {
-            source.closed.catch(abortDest);
-          } else if (source.state === 'closed') {
-            closeDest();
-          }
-        } else if (ds === 'waiting') {
-          if (source.state === 'readable') {
-            Promise.race([source.closed, dest.ready]).then(doPipe, doPipe);
-          } else if (source.state === 'waiting') {
-            Promise.race([source.ready, dest.ready]).then(doPipe);
-          } else if (source.state === 'errored') {
-            source.closed.catch(abortDest);
-          } else if (source.state === 'closed') {
-            closeDest();
-          }
-        } else if (ds === 'errored' && (source.state === 'readable' || source.state === 'waiting')) {
-          dest.closed.catch(cancelSource);
-        } else if ((ds === 'closing' || ds === 'closed') &&
-            (source.state === 'readable' || source.state === 'waiting')) {
-          cancelSource(new TypeError('destination is closing or closed and cannot be piped to anymore'));
+      Promise.all([source.read(), dest.ready]).then(([chunk]) => {
```

Comment:
Hmm. I guess in that case we will just keep doing `dest.write()` and ignoring the fact that it fails. That is indeed silly and so we should not do it. But it at least explains why no tests were failing.